src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
+        } else {
+            addNodeToEnd(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (size == 0) {
+            addFirstNode(value);
+        } else if (index == 0) {
+            addToTop(value);
+        } else if (index == size) {
+            addNodeToEnd(value);
+        } else {
+            Node<T> currentNode = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(value, currentNode, currentNode.prev);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.element;
+        currentNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        return unlink(node);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if ((object == null && currentNode.element == null) || (object != null && object.equals(
+                    currentNode.element))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addFirstNode(T element) {
+        first = new Node<>(element, null, null);
+        last = first;
+        size++;
+    }
+
+    private void addToTop(T value) {
+        Node<T> newNode = new Node<>(value, first, null);
+        first.prev = newNode;
+        first = newNode;
+        size++;
+    }
+
+    private void addNodeToEnd(T value) {
+        Node<T> newNode = new Node<>(value, null, last);
+        last.next = newNode;
+        last = newNode;
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
```

Comment:
Don't iterate your nodes from the head to the tail or vice versa. Based on the index, you should iterate only the first or second half of the list to improve performance.